README.md

# NanoPcap
NanoPcap is a Python library and set of tools for working with nanosecond
resolution PCAP data. It is designed to be minimal and require no dependencies.

## Tools

### `Dump`
Dumps a PCAP in either short form (1 line per packet) or long form (1 line per
value).

	> NanoPcap/Tools/Dump.py -h
	usage: Dump.py [-h] [-d DATA_BYTES] [-l] [-j] [-o DATA_OFFSET] [-H] [-R] [-s]
	               pcap

	PCAP Dump Diagnostic

	positional arguments:
	  pcap                  PCAP file to dump.

	optional arguments:
	  -h, --help            show this help message and exit
	  -d DATA_BYTES, --data-bytes DATA_BYTES
	                        Show a certain number of bytes as hex for each packet
	                        record.
	  -l, --long            Enable long form which generally puts one value per
	                        line for easy diffing.
	  -j, --json            Enable JSON output with either one object per line
	                        (short mode) or one value per line (long mode).
	  -o DATA_OFFSET, --data-offset DATA_OFFSET
	                        Offset of the data to show.
	  -H, --no-header       Do not show the header.
	  -R, --no-records      Do not show records.
	  -s, --strict          Enables strict validation rules.

### `Filter`
Filters a PCAP based on set criteria and optionally does other edits like snapshot
length truncation, packet deduplication, or even fuzzing like random drops and duplication.

	> NanoPcap/Tools/Filter.py -h
	usage: Filter.py [-h] [--strict] [-l SNAPLEN] [-o DATA_OFFSET]
	                 [-x DATA_END_OFFSET] [-H] [-R] [-a]
	                 [--required-link-type REQUIRED_LINK_TYPE]
	                 [--link-type LINK_TYPE]
	                 [--time-shift-seconds TIME_SHIFT_SECONDS] [-s START] [-e END]
	                 [-D DROP_FRACTION] [--duplicate-fraction DUPLICATE_FRACTION]
	                 [--deduplication-window DEDUPLICATION_WINDOW]
	                 input output

	PCAP Filter Tool

	positional arguments:
	  input                 PCAP file to use as input.
	  output                Output file. May include time format strings to roll
	                        the file based on packet time stamps, e.g.
	                        %Y/%m/%d/%H.pcap for hourly output files in daily
	                        folders.

	optional arguments:
	  -h, --help            show this help message and exit
	  --strict              Enables strict validation rules.
	  -l SNAPLEN, --snaplen SNAPLEN
	                        Add a certain number of bytes for each packet record.
	  -o DATA_OFFSET, --data-offset DATA_OFFSET
	                        Offset of the data to include.
	  -x DATA_END_OFFSET, --data-end-offset DATA_END_OFFSET
	                        Offset from the end of the data to include.
	  -H, --no-header       Do not output the header.
	  -R, --no-records      Do not output records.
	  -a, --append          Append to the file (implies no header).
	  --required-link-type REQUIRED_LINK_TYPE
	                        The required link type of the file being edited (e.g.
	                        1 for Ethernet, 228 for IPv4, 229 for IPv6).
	  --link-type LINK_TYPE
	                        A value to set the link type in the header to (e.g. 1
	                        for Ethernet, 228 for IPv4, 229 for IPv6).
	  --time-shift-seconds TIME_SHIFT_SECONDS
	                        The amount of time in seconds to shift timestamps in
	                        the output PCAP.
	  -s START, --start START
	                        Start time as either epoch nanoseconds or a datetime
	                        (with only microsecond resolution).
	  -e END, --end END     End time as either epoch nanoseconds or a relative
	                        offset in nanoseconds to the start (e.g. +100 would
	                        yield a 100ns PCAP).
	  -D DROP_FRACTION, --drop-fraction DROP_FRACTION
	                        Fraction of the time to drop packagets (from 0 to 1
	                        inclusive).
	  --duplicate-fraction DUPLICATE_FRACTION
	                        Fraction of the time to duplicate packagets (from 0 to
	                        1 inclusive).
	  --deduplication-window DEDUPLICATION_WINDOW
	                        Sets the number of the packets in the deduplication
	                        window (based on contents).

For example, here is how Ethernet headers (L2) were removed to generate the files in TestData:

	> NanoPcap/Tools/Filter.py --required-link-type 1 --link-type 228 -o 14 -x 4 SSH.pcap TestData/SSH_L3.pcap
	> NanoPcap/Tools/Filter.py --required-link-type 1 --link-type 228 -o 14 -x 4 SSH2.pcap TestData/SSH2_L3.pcap

There is also a convenience script for that transformation:

	> ./strip_ethernet_header.sh SSH.pcap TestData/SSH_L3.pcap

### `Merge`
Merges two PCAP files with potentially interleaved timestamps.

	> NanoPcap/Tools/Merge.py -h
	usage: Merge.py [-h] [--strict] input1 input2 output

	PCAP Filter Tool

	positional arguments:
	  input1      PCAP file to use as input.
	  input2      PCAP file to use as other input.
	  output      Output file

	optional arguments:
	  -h, --help  show this help message and exit
	  --strict    Enables strict validation rules.

### `Split`
Splits a PCAP into slices with a maximum number of packets, bytes, etc.

	> NanoPcap/Tools/Split.py -h
	usage: Split.py [-h] [--gzip-output] [--strict] [-b MAX_BYTES]
	                [-p MAX_PACKETS] [-l SNAPLEN] [-o DATA_OFFSET]
	                [-x DATA_END_OFFSET] [-H] [-a]
	                input output

	PCAP Splitting Tool

	positional arguments:
	  input                 PCAP file to use as input.
	  output                Output path -- output files will be named based on the
	                        identifying attributes.

	optional arguments:
	  -h, --help            show this help message and exit
	  --gzip-output         Enables gzip for the output files.
	  --strict              Enables strict validation rules.
	  -b MAX_BYTES, --max-bytes MAX_BYTES
	                        The maximum number of bytes in a slice.
	  -p MAX_PACKETS, --max-packets MAX_PACKETS
	                        The maximum number of packets in a slice.
	  -l SNAPLEN, --snaplen SNAPLEN
	                        Add a certain number of bytes for each packet record.
	  -o DATA_OFFSET, --data-offset DATA_OFFSET
	                        Offset of the data to include.
	  -x DATA_END_OFFSET, --data-end-offset DATA_END_OFFSET
	                        Offset from the end of the data to include.
	  -H, --no-header       Do not output the header.
	  -a, --append          Append to the file (implies no header).

### `SplitFlows`
Splits a PCAP into multiple PCAP's, one per flow at the top layer protocol.

	> mkdir -p SplitData && NanoPcap/Tools/SplitEthernetFlows.py TestData/SSH_L3.pcap SplitData/ && ls SplitData/
	192.168.1.192_192.168.1.241.pcap

	> NanoPcap/Tools/SplitFlows.py -h
	usage: SplitFlows.py [-h] [--strict] [-l SNAPLEN] [-o DATA_OFFSET]
	                     [-x DATA_END_OFFSET] [-H] [-a] [--link-type LINK_TYPE]
	                     input output

	PCAP Filter Tool

	positional arguments:
	  input                 PCAP file to use as input.
	  output                Output path -- output files will be named based on the
	                        identifying attributes.

	optional arguments:
	  -h, --help            show this help message and exit
	  --strict              Enables strict validation rules.
	  -l SNAPLEN, --snaplen SNAPLEN
	                        Add a certain number of bytes for each packet record.
	  -o DATA_OFFSET, --data-offset DATA_OFFSET
	                        Offset of the data to include.
	  -x DATA_END_OFFSET, --data-end-offset DATA_END_OFFSET
	                        Offset from the end of the data to include.
	  -H, --no-header       Do not output the header.
	  -a, --append          Append to the file (implies no header).
	  --link-type LINK_TYPE
	                        A value to set the link type in the header to (e.g. 1
	                        for Ethernet, 228 for IPv4, 229 for IPv6).

### `Summary`
Summarizes a PCAP. For example:

	> NanoPcap/Tools/Summary.py TestData/SSH_L3.pcap -u
	Epoch times: 1472402096321502000 - 1472402096321652000 (150000ns) (2016-08-28 16:34:56.321501 - 2016-08-28 16:34:56.321651)

	Name                      Count    Total  Average  Std Dev      Min   25th %   50th %   75th %   95th %   99th % 99.9th %      Max
	Included Length              21     8.6K   421.43   502.94       34       34      102      582     1482     1482     1482     1482
	Original Length              21     8.6K   421.43   502.94       34       34      102      582     1482     1482     1482     1482
	Interpacket Time (ns)        20  150.0us    7.5us   20.9us      0.0      0.0    1.0us    1.0us   74.0us   74.0us   74.0us   74.0us
	Packet Rate (pps)            20            133.3K             13.5K     1.0M      inf      inf      inf      inf      inf      inf
	Data Rate (Bps)              20                              448.7K   539.8M      inf      inf      inf      inf      inf      inf

Or without units:

	> NanoPcap/Tools/Summary.py TestData/SSH_L3.pcap
	Epoch times: 1472402096321502000 - 1472402096321652000 (150000ns) (2016-08-28 16:34:56.321501 - 2016-08-28 16:34:56.321651)

	Name                          Count            Total        Average        Std Dev            Min         25th %         50th %         75th %         95th %         99th %       99.9th %            Max
	Included Length                  21           8850.0         421.43         502.94             34             34            102            582           1482           1482           1482           1482
	Original Length                  21           8850.0         421.43         502.94             34             34            102            582           1482           1482           1482           1482
	Interpacket Time (ns)            20         150000.0         7500.0        20884.2            0.0            0.0         1000.0         1000.0        74000.0        74000.0        74000.0        74000.0
	Packet Rate (pps)                20                        133333.3                       13513.5      1000000.0            inf            inf            inf            inf            inf            inf
	Data Rate (Bps)                  20                                                      459459.5    566000000.0            inf            inf            inf            inf            inf            inf


CHANGELOG.md
# Change Log

## [1.0.4]
### Changed
- Raise ValueError on invalid PCAP magic in strict mode.

## [1.0.3] - 2022-11-27
### Removed
- Removed obsolete documentation generation.

## [1.0.2] - 2022-11-27
### Changed
- Fix imports so tools can be used directly.

## [1.0.1] - 2021-04-21
### Changed
- Fix missing submodules in PyPi distribution.

## [1.0.0] - 2020-11-15
### Added
- Optionally force matching link types in the merge tool.
- Smoke tests for `Split` and `SplitFlows` tools
### Changed
- Always write headers to empty files, even in append mode.

## [0.1.0] - 2020-04-13
### Added
- This change log.
### Changed
- Fix `Filter` tool append mode file header generation when appending to an empty file.
- Clean up typos in documentation and error messages.

[1.0.4]: https://github.com/AgalmicVentures/HumanTime/compare/1.0.3...1.0.4
[1.0.3]: https://github.com/AgalmicVentures/HumanTime/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/AgalmicVentures/HumanTime/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/AgalmicVentures/HumanTime/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/AgalmicVentures/HumanTime/compare/0.1.0...1.0.0
[0.1.0]: https://github.com/AgalmicVentures/HumanTime/compare/0.0.1...0.1.0
[0.0.1]: https://github.com/AgalmicVentures/HumanTime/releases/tag/0.0.1


NanoPcap/Tools/Filter.py
#!/usr/bin/env python3

# Copyright (c) 2015-2022 Agalmic Ventures LLC (www.agalmicventures.com)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import argparse
import datetime
import gzip
import os
import random
import sys

import inspect
_currentFile = os.path.abspath(inspect.getfile(inspect.currentframe()))
_currentDir = os.path.dirname(_currentFile)
_parentDir = os.path.dirname(os.path.dirname(_currentDir))
sys.path.insert(0, _parentDir)

from NanoPcap import Listener, Parser
from NanoPcap.Utility import Data

class PcapFilterListener(Listener.PcapListener):

	def __init__(self, arguments):
		self._arguments = arguments

		self._lastPacketDatas = []
		self._outputFileName = None
		self._outputFile = None
		self._header = None

	def onPcapHeader(self, header):
		self._header = header

		#Warn on a larger snap len than before
		if self._arguments.snaplen > header.snaplen():
			print('WARNING: New snaplen is greater than original: %d > %d' % (self._arguments.snaplen, header.snaplen()))

		#Check if the link type is correct
		if self._arguments.required_link_type is not None and self._arguments.required_link_type != header.network():
			print('ERROR: Link type is %d instead of %d' % (header.network(), self._arguments.required_link_type))
			sys.exit(1)

		#Update with new snaplen
		snaplen = min(self._arguments.snaplen, header.snaplen())
		self._header.setSnaplen(snaplen)

		#Update with the new link link
		if self._arguments.link_type is not None:
			self._header.setNetwork(self._arguments.link_type)

	def onPcapRecord(self, recordHeader, data):
		if self._arguments.no_records:
			return

		#Filter

		#Check the time
		epochTime = recordHeader.epochNanos()
		if self._arguments.start is not None and epochTime < self._arguments.start:
			return
		if self._arguments.end is not None and epochTime > self._arguments.end:
			return

		#Drop?
		if self._arguments.drop_fraction > 0 and random.random() < self._arguments.drop_fraction:
			return

		#De-duplicate
		if self._arguments.deduplication_window > 0:
			found = False
			for packetData in self._lastPacketDatas:
				if packetData == data:
					found = True
					break

			self._lastPacketDatas.append(data)
			if len(self._lastPacketDatas) > self._arguments.deduplication_window:
				self._lastPacketDatas.pop(0)
			if found:
				return

		#Roll the file if necessary
		newOutputFileName = recordHeader.timestampDatetime().strftime(self._arguments.output)
		if newOutputFileName != self._outputFileName:
			if self._outputFile is not None:
				self._outputFile.close()

			#Write to the output file
			self._outputFileName = newOutputFileName
			directory, fileName = os.path.split(self._outputFileName)
			if directory != '' and not os.path.exists(directory):
				os.makedirs(directory)

			mode = 'ab' if self._arguments.append else 'wb'
			self._outputFile = gzip.open(self._outputFileName, mode) if self._outputFileName.endswith('.gz') else open(self._outputFileName, mode)
			#Write the header at the beginning, unless instructed otherwise
			#This neatly handles append mode
			if not self._arguments.no_header and self._outputFile.tell() == 0:
				self._header.writeToFile(self._outputFile)

		#Update with new snaplen
		start = self._arguments.data_offset
		end = min(start + self._arguments.snaplen, len(data) - start - self._arguments.data_end_offset)
		truncatedData = data[start:end]
		recordHeader.setIncludedLength(len(truncatedData))
		if self._arguments.link_type is not None:
			recordHeader.setOriginalLength(len(truncatedData))

		#Update with the new timestamp if necessary
		if self._arguments.time_shift_seconds is not None:
			recordHeader.setTsSec(recordHeader.tsSec() + self._arguments.time_shift_seconds)

		#Write the header
		recordHeader.writeToFile(self._outputFile)

		#Write the data, randomizing if necessary
		if self._arguments.data_randomization_fraction > 0:
			newTruncatedData = Data.randomizeBytes(truncatedData, self._arguments.data_randomization_fraction)
			self._outputFile.write(newTruncatedData)
		else:
			self._outputFile.write(truncatedData)

		#Duplicate?
		if self._arguments.duplicate_fraction > 0 and random.random() < self._arguments.duplicate_fraction:
			recordHeader.writeToFile(self._outputFile)
			self._outputFile.write(truncatedData)

def datetimeToEpochNanos(dt):
	seconds = (dt - datetime.datetime(1970, 1, 1)).total_seconds()
	return seconds * 1000 * 1000 * 1000

def main():
	parser = argparse.ArgumentParser(description='PCAP Filter Tool')
	parser.add_argument('input', help='PCAP file to use as input.')
	parser.add_argument('output', help='Output file. May include time format strings to roll the file based on packet time stamps, e.g. %%Y/%%m/%%d/%%H.pcap for hourly output files in daily folders.')

	#Validation
	parser.add_argument('--strict', action='store_true',
		help='Enables strict validation rules.')

	#"Where to cut" in bytes
	parser.add_argument('-l', '--snaplen', type=int, default=65535, action='store',
		help='Add a certain number of bytes for each packet record.')
	parser.add_argument('-o', '--data-offset', type=int, default=0, action='store',
		help='Offset of the data to include.')
	parser.add_argument('-x', '--data-end-offset', type=int, default=0, action='store',
		help='Offset from the end of the data to include.')
	parser.add_argument('-H', '--no-header', action='store_true',
		help='Do not output the header.')
	parser.add_argument('-R', '--no-records', action='store_true',
		help='Do not output records.')
	parser.add_argument('-a', '--append', action='store_true',
		help='Append to the file (implies no header).')
	parser.add_argument('--data-randomization-fraction', type=float, default=0.0, action='store',
		help='Fraction of the data bytes to randomize (from 0 to 1 inclusive).')

	#Header edits
	parser.add_argument('--required-link-type', type=int, default=None, action='store',
		help='The required link type of the file being edited (e.g. 1 for Ethernet, 228 for IPv4, 229 for IPv6).')
	parser.add_argument('--link-type', type=int, default=None, action='store',
		help='A value to set the link type in the header to (e.g. 1 for Ethernet, 228 for IPv4, 229 for IPv6).')

	parser.add_argument('--time-shift-seconds', type=int, default=None, action='store',
		help='The amount of time in seconds to shift timestamps in the output PCAP.')

	#Filtering
	parser.add_argument('-s', '--start', default=None, action='store',
		help='Start time as either epoch nanoseconds or a datetime (with only microsecond resolution).')
	parser.add_argument('-e', '--end', default=None, action='store',
		help='End time as either epoch nanoseconds or a relative offset in nanoseconds to the start (e.g. +100 would yield a 100ns PCAP).')

	parser.add_argument('-D', '--drop-fraction', type=float, default=0.0, action='store',
		help='Fraction of the time to drop packagets (from 0 to 1 inclusive).')
	parser.add_argument('--duplicate-fraction', type=float, default=0.0, action='store',
		help='Fraction of the time to duplicate packagets (from 0 to 1 inclusive).')
	parser.add_argument('--deduplication-window', type=int, default=0, action='store',
		help='Sets the number of the packets in the deduplication window (based on contents).')

	arguments = parser.parse_args(sys.argv[1:])

	#Parse the start time
	if arguments.start is not None:
		try:
			arguments.start = int(arguments.start)
		except ValueError:
			arguments.start = datetimeToEpochNanos(datetime.datetime.strptime(arguments.start, '%Y-%m-%d %H:%M:%S.%f'))

	#Parse the end time
	if arguments.end is not None:
		try:
			relativeEnd = arguments.end[0] == '+' and arguments.start is not None
			arguments.end = int(arguments.end)
			if relativeEnd:
				arguments.end += arguments.start
		except ValueError:
			arguments.end = datetimeToEpochNanos(datetime.datetime.strptime(arguments.end, '%Y-%m-%d %H:%M:%S.%f'))

	listener = PcapFilterListener(arguments)
	Parser.parseFile(arguments.input, listener, strict=arguments.strict)

	return 0

if __name__ == '__main__':
	sys.exit(main())
